topology/nhlt/intel/dmic-nhlt.c

```diff
@@ -0,0 +1,524 @@
+	struct mic_vendor_config mic_conf;
+	struct formats_config f_conf;
+	struct format_config f_conf1;
+	uint8_t *ep_target;
+	size_t blob_size;
+	int ret;
+	int i;
+
+	size_t mic_config_size;
+	uint32_t sample_rate;
+	uint16_t channel_count;
+	uint32_t bits_per_sample;
+	uint8_t array_type;
+	uint8_t extension;
+	uint8_t num_mics;
+	uint32_t snr;
+	uint32_t sensitivity;
+
+	uint8_t type;
+	uint8_t panel;
+	uint32_t speaker_position_distance;
+	uint32_t horizontal_offset;
+	uint32_t vertical_offset;
+	uint8_t frequency_low_band;
+	uint8_t frequency_high_band;
+	uint16_t direction_angle;
+	uint16_t elevation_angle;
+	uint16_t vertical_angle_begin;
+	uint16_t vertical_angle_end;
+	uint16_t horizontal_angle_begin;
+	uint16_t horizontal_angle_end;
+
+	/*
+	 * nhlt dmic structure:
+	 *
+	 * endpoint_descriptor, sizeof(struct endpoint_descriptor)
+	 *
+	 * device_specific_config (mic), sizeof(mic_array_device_specific_config)
+	 * or
+	 * device_specific_config (mic), sizeof(mic_array_device_specific_vendor_config)
+	 *
+	 * formats_config (formats_count), sizeof(struct formats_config)
+	 * format_config (waveex), sizeof(struct format_config)
+	 * vendor_blob sizeof(vendor_blob)
+	 */
+
+	/* dmic ep */
+	ep.link_type = NHLT_LINK_TYPE_PDM;
+	ep.instance_id = 0;
+	ep.vendor_id = NHLT_VENDOR_ID_INTEL;
+	ep.device_id = NHLT_DEVICE_ID_INTEL_PDM_DMIC;
+	ep.revision_id = 0;
+	ep.subsystem_id = 0;
+	ep.device_type = 0;
+	ep.direction = NHLT_ENDPOINT_DIRECTION_CAPTURE;
+	ep.virtualbus_id = 0;
+
+	ret = dmic_get_params(nhlt, index, &sample_rate, &channel_count, &bits_per_sample,
+			      &array_type, &num_mics, &extension, &snr, &sensitivity);
+
+	if (ret) {
+		fprintf(stderr, "nhlt_dmic_get_ep: dmic_get_params failed\n");
+		return ret;
+	}
+
+	if (array_type == NHLT_MIC_ARRAY_TYPE_VENDOR_DEFINED) {
+		mic_v_conf.config.capabilities_size = 4 + num_mics *
+			sizeof(struct mic_vendor_config);
+		mic_v_conf.device_config.virtual_slot = 0; /* always 0 for dmic */
+		mic_v_conf.device_config.config_type = NHLT_DEVICE_CONFIG_TYPE_MICARRAY;
+		mic_v_conf.number_of_microphones = num_mics;
+		mic_v_conf.array_type_ex = array_type;
+		/* presense of extension struct is coded into lower 4 bits of array_type */
+		if (extension) {
+			mic_v_conf.array_type_ex = (array_type & ~0x0F) | (0x01 & 0x0F);
+			mic_v_conf.config.capabilities_size +=
+				sizeof(struct mic_snr_sensitivity_extension);
+		}
+	} else {
+		mic_s_conf.config.capabilities_size = 3;
+		mic_s_conf.device_config.virtual_slot = 0; /* always 0 for dmic */
+		mic_s_conf.device_config.config_type = NHLT_DEVICE_CONFIG_TYPE_MICARRAY;
+		mic_s_conf.array_type_ex = array_type;
+		/* presense of extension struct coded into lower 4 bits of array_type */
+		if (extension) {
+			mic_s_conf.array_type_ex = (array_type & ~0x0F) | (0x01 & 0x0F);
+			mic_s_conf.config.capabilities_size +=
+				sizeof(struct mic_snr_sensitivity_extension);
+		}
+	}
+
+	/* formats_config */
+	f_conf.formats_count = 1;
```

Comment:
we can, but our dmic topology hasn't ever supported multiple configs. It has always been 1 single config. I see some nhlt blobs contain the 16 and 32 configs separately, but seems that 32 bit is the most common these day. So to summarize, if we make the support for multiple configs in topology, this can be added.

I'm not sure what kind of variations would be generally in bios nhlt for dmic, would be good info.